homeassistant/components/keymitt_ble/config_flow.py

```diff
@@ -0,0 +1,154 @@
+            discovery_info.device, discovery_info.advertisement
+        )
+        self._discovered_adv = parsed
+        self.context["title_placeholders"] = {
+            "name": name_from_discovery(self._discovered_adv),
+        }
+        return await self.async_step_init()
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initialized by the user."""
+        # This is for backwards compatibility.
+        return await self.async_step_init(user_input)
+
+    async def async_step_init(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Check if paired."""
+        errors: dict[str, str] = {}
+
+        if discovery := self._discovered_adv:
+            self._discovered_advs[discovery.address] = discovery
+        else:
+            current_addresses = self._async_current_ids()
+            for discovery_info in async_discovered_service_info(self.hass):
+                self._ble_device = discovery_info.device
+                address = discovery_info.address
+                if (
+                    address in current_addresses
+                    or address in self._discovered_advs
+                ):
+                    continue
+                parsed = parse_advertisement_data(
+                    discovery_info.device, discovery_info.advertisement
+                )
+                if parsed:
+                    self._discovered_advs[address] = parsed
+
+        if not self._discovered_advs:
+            return self.async_abort(reason="no_unconfigured_devices")
+
+        if user_input is not None:
+            self._name = name_from_discovery(self._discovered_adv)
+            self._bdaddr = user_input[CONF_ADDRESS]
+            await self.async_set_unique_id(
+                self._bdaddr, raise_on_progress=False
+            )
+            self._abort_if_unique_id_configured()
+            return await self.async_step_link()
+
+        return self.async_show_form(
+            step_id="init",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_ADDRESS): vol.In(
+                        {
+                            address: f"{parsed.data['local_name']} ({address})"
+                            for address, parsed in self._discovered_advs.items()
+                        }
+                    )
+                }
+            ),
+            errors=errors,
+        )
+
+    async def async_step_link(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Given a configured host, will ask the user to press the button to pair."""
+        errors: dict[str, str] = {}
+        token = randomid(32)
+        self._client = MicroBotApiClient(
+            device=self._ble_device,
+            token=token,
+        )
+        assert self._client is not None
+        if user_input is None:
```

Comment:
If you reverse this condition you can eliminate the duplicate `return self.async_show_form(step_id="link")` by having it fall through to the end if there is an error or create if there is not

Reply:
I'll have to play around with that. I want the form to show right after the token is sent to the device, and I want it to show again if unsuccessful

Reply:
I don't have the device so I can't know for sure, but I think you could re-write it as
```python
    async def async_step_link(
        self, user_input: dict[str, Any] | None = None
    ) -> FlowResult:
        """Given a configured host, will ask the user to press the button to pair."""
        errors: dict[str, str] = {}
        assert self._bdaddr is not None

        if user_input is not None:
            token = randomid(32)
            client = MicroBotApiClient(
                device=self._ble_device,
                token=token,
            )            
            await client.connect(init=True)
            if client.is_connected():
                await client.disconnect()
                return self.async_create_entry(title=self._name, data={
                    CONF_ACCESS_TOKEN: token,
                    CONF_ADDRESS: self._bdaddr,
                }) 
            errors["base"] = "linking"

        return self.async_show_form(step_id="link", errors=errors)
```

Reply:
The problem with doing it this way is that I would need to put in a wait, or a while true, or something, as I'm waiting for the user to press the button on the bot. This errors out too quickly. 

Actually the above creates the device too quickly as is_connected is true before the bot button is pressed. I've replaced with if client.token_callback == token: but this errors straight away as the user interaction hasn't happened yet.

I can persist if you think it's necessary but the original code works as the user clicks submit when done.

Reply:
Its ok to leave it as-is.  Do we need the disconnect though?

Reply:
Well no because I had it in the library.....but I can see a potential problem with that... so yes I have moved it here